### ISSUE TITLE
Model Checkpoints

### ISSUE THREAD

**Issue by Andrea94c** [vanderschaarlab/synthcity#133]
## Question
Is it possible to create checkpoints during training? And/or resume training from a previously saved model?

**Comment by bcebere** [comment]
Hello @Andrea94c 

Thank you for the question.

We haven't added a model saving support mainly because the library is still in development, and the internals can frequently change, invalidating the previous caches.

However, we will investigate how to safely cache checkpoints. For which models would you want this kind of support?

Thank you!

**Comment by pravsels** [comment]
Hey @bcebere

I've been doing this manually, training for 50 epochs or so, saving the model to disk, loading it back from disk and continue training. 

But I'd love to pass in an argument like `n_iter_save` and have checkpoints saved every few iterations. Or maybe have it linked to `n_iter_print`, that's how I assumed it worked initially. I've been trying out GANs, so having that option for it would be nice. 

This would presumably go in the /synthcity/plugins/core/models/gan.py file ? What do you think about me submitting a PR for this ?

**Comment by Andrea94c** [comment]
Thanks for the answers, I'm working with GANs as well.

**Comment by bcebere** [comment]
@pravsels Thank you for your feedback.

Any contribution is more than welcome.

However, caching the models can be tricky. More exactly, any change in the models will break the existing caches.
So, the mechanism should consider the library's version, architecture/OS, and python version.

If you want to tackle this, I can provide feedback on the PR. Otherwise, I will have a look in the following weeks.

Many thanks!

cc @robsdavis

**Comment by pravsels** [comment]
> @pravsels Thank you for your feedback.
> 
> Any contribution is more than welcome.
> 
> However, caching the models can be tricky. More exactly, any change in the models will break the existing caches. So, the mechanism should consider the library's version, architecture/OS, and python version.
> 
> If you want to tackle this, I can provide feedback on the PR. Otherwise, I will have a look in the following weeks.
> 
> Many thanks!
> 
> cc @robsdavis

Yes, I would like to take a crack at it. 

Would appreciate any feedback you can provide to help me in the process, maybe even a link to another PR that might've tackled something similar.

**Comment by bcebere** [comment]
@pravsels Thank you very much for working on this.

However, I would suggest to wait a few days before starting, we have a big merge with the image generation support, and most likely will create a ton of merge conflicts in your branch.

After that, there are a few things to keep in mind:
 - All the plugins have a [`workspace`](https://github.com/vanderschaarlab/synthcity/blob/main/src/synthcity/plugins/generic/plugin_ctgan.py#L136) parameter that needs now to be propagated -> TabularGAN -> GAN, so you save in the same folder. 
 - When there is a `patience_metric` in the core models, there is a logic for saving the best model version [here](https://github.com/vanderschaarlab/synthcity/blob/main/src/synthcity/plugins/core/models/gan.py#L636).
 - You would need to see if you want to save `best_state_dict` or just the latest, if you have or not the `patience_metric`.
 - When you save `best_state_dict`/`latest_state_dict` to a file, that file should be in the `workspace` path, and with a name that includes
   -  the synthcity [version`](https://github.com/vanderschaarlab/synthcity/blob/main/src/synthcity/version.py). If there is a checkpoint or backup from a different synthcity version, it is best to just ignore that, to prevent errors and crashes.
   - the python version, in case you have multiple conda envs.
   - a hash of the training set.

Hopefully, these 3 identifiers(synthcity version, python version, dataset hash), are enough to avoid errors in the cached models.

Thank you again for your contribution! Looking forward to the PR. :)
